src/Middleware/MiddlewareInterface.php

```diff
@@ -0,0 +1,49 @@
+<?php
+/*
+ * Copyright 2023 Google LLC
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+namespace Google\ApiCore\Middleware;
+
+use Google\ApiCore\Call;
+use GuzzleHttp\Promise\PromiseInterface;
+use Google\ApiCore\ClientStream;
+use Google\ApiCore\ServerStream;
+use Google\ApiCore\BidiStream;
+
+interface MiddlewareInterface
```

Comment:
I think the expectation of implementors is that they invoke the next middleware, is that correct? If so, I think we need some more explicit documentation about this. I don't see it in here that "a MiddlewareInterface implementation **must** call the MiddlewareInterface it is given". Not sure where it goes though.

Reply:
Yes you are correct!  I struggled with this, because right now the handler is passed in the constructor (e.g. `__construct(MiddlewareInterface $nextHandler)`) and then called in `__invoke`, which is noted by the `PromiseInterface` typehint. 

The issue I ran into is: 
1. The constructors can (and should) differ. So I can't put that in the interface. 
2. The return types can _also_ be `BidiStream`, `ClientStream`, or `ServerStream`, so we can't typehint that either.

Would it be sufficient to expand the documentation to explain these things? I'll go ahead and make that change now.

Reply:
See https://github.com/googleapis/gax-php/pull/473/commits/75a54947f82958ea875b37b09740f87a38b6b21f

Reply:
SGTM thanks